mr/src/main/java/org/apache/iceberg/mr/Catalogs.java

```diff
@@ -180,47 +189,97 @@ public static boolean dropTable(Configuration conf, Properties props) {
-  public static boolean hiveCatalog(Configuration conf) {
-    return HIVE.equalsIgnoreCase(conf.get(InputFormatConfig.CATALOG));
+  public static boolean hiveCatalog(Configuration conf, Properties props) {
+    String catalogName = props.getProperty(InputFormatConfig.TABLE_CATALOG);
+    if (catalogName != null) {
+      return HIVE.equalsIgnoreCase(conf.get(String.format(InputFormatConfig.CATALOG_TYPE_TEMPLATE, catalogName)));
+    } else {
+      if (HIVE.equalsIgnoreCase(conf.get(String.format(InputFormatConfig.CATALOG_TYPE_TEMPLATE, DEFAULT_CATALOG)))) {
+        return true;
+      } else {
+        return HIVE.equalsIgnoreCase(conf.get(InputFormatConfig.CATALOG));
+      }
+    }
   }
 
   @VisibleForTesting
-  static Optional<Catalog> loadCatalog(Configuration conf) {
-    String catalogLoaderClass = conf.get(InputFormatConfig.CATALOG_LOADER_CLASS);
-
-    if (catalogLoaderClass != null) {
-      CatalogLoader loader = (CatalogLoader) DynConstructors.builder(CatalogLoader.class)
-              .impl(catalogLoaderClass)
-              .build()
-              .newInstance();
-      Catalog catalog = loader.load(conf);
-      LOG.info("Loaded catalog {} using {}", catalog, catalogLoaderClass);
-      return Optional.of(catalog);
+  static Optional<Catalog> loadCatalog(Configuration conf, String catalogName) {
+    String catalogType;
+    String name = catalogName == null ? DEFAULT_CATALOG : catalogName;
+    catalogType = conf.get(String.format(InputFormatConfig.CATALOG_TYPE_TEMPLATE, name));
+
+    // keep both catalog configuration methods for seamless transition
+    if (catalogType != null) {
+      // new logic
+      LOG.debug("Using catalog configuration from table properties.");
+      return loadCatalog(conf, name, catalogType);
+    } else {
+      // old logic
+      // use catalog {@link InputFormatConfig.CATALOG} stored in global hive config if table specific catalog
+      // configuration or default catalog definition is missing
+      LOG.debug("Using catalog configuration from global configuration.");
+      catalogType = conf.get(InputFormatConfig.CATALOG);
+      return loadCatalog(conf, name, catalogType);
     }
+  }
 
-    String catalogName = conf.get(InputFormatConfig.CATALOG);
+  private static Optional<Catalog> loadCatalog(Configuration conf, String catalogName, String catalogType) {
+    if (catalogType == null) {
+      LOG.info("Catalog is not configured");
+      return Optional.empty();
+    }
 
-    if (catalogName != null) {
-      Catalog catalog;
-      switch (catalogName.toLowerCase()) {
-        case HADOOP:
+    Map<String, String> properties = getCatalogProperties(conf, catalogName);
+    Catalog catalog;
+    switch (catalogType.toLowerCase()) {
+      case HADOOP:
+        if (properties.containsKey(CatalogProperties.WAREHOUSE_LOCATION)) {
+          catalog = CatalogUtil.loadCatalog(HadoopCatalog.class.getName(), catalogName,
+                  getCatalogProperties(conf, catalogName), conf);
```

Comment:
We do have this properties at hand already? Don't we?

Reply:
Right, I missed that.